tests/test_interpolation.py

```diff
@@ -753,3 +756,250 @@ def test_none_value_in_quoted_string(restore_resolvers: Any) -> None:
+                        """\
+                        Value 'Bond' could not be converted to Integer
+                            full_key: age
+                        """
+                    )
+                ),
+            ),
+            id="type_mismatch_node_interpolation",
+        ),
+        pytest.param(
+            StructuredWithMissing(opt_num=None, num=II("opt_num")),
+            "num",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape("Non optional field cannot be assigned None"),
+            ),
+            id="non_optional_node_interpolation",
+        ),
+        pytest.param(
+            SubscriptedList(list=SI("${identity:[a, b]}")),
+            "list",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape("Value 'a' could not be converted to Integer"),
+            ),
+            id="list_type_mismatch",
+        ),
+        pytest.param(
+            MissingDict(dict=SI("${identity:{0: b, 1: d}}")),
+            "dict",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape("Key 0 (int) is incompatible with (str)"),
+            ),
+            id="dict_key_type_mismatch",
+        ),
+    ],
+)
+def test_interpolation_type_validated_error(
+    cfg: Any,
+    key: str,
+    expected_error: Any,
+    register_identity: Any,
+) -> None:
+    def cast(t: Any, v: Any) -> Any:
+        return {"str": str, "int": int}[t](v)  # cast `v` to type `t`
+
+    OmegaConf.register_new_resolver("cast", cast)
+
+    cfg = OmegaConf.structured(cfg)
+
+    with expected_error:
+        cfg[key]
+
+    assert OmegaConf.select(cfg, key, throw_on_resolution_failure=False) is None
+
+
+@pytest.mark.parametrize(
+    ("cfg", "key"),
+    [
+        pytest.param({"dict": "${identity:{a: 0, b: 1}}"}, "dict.a", id="dict"),
+        pytest.param(
+            {"dict": "${identity:{a: 0, b: {c: 1}}}"},
+            "dict.b.c",
+            id="dict_nested",
+        ),
+        pytest.param({"list": "${identity:[0, 1]}"}, "list.0", id="list"),
+        pytest.param({"list": "${identity:[0, [1, 2]]}"}, "list.1.1", id="list_nested"),
+    ],
+)
+def test_interpolation_readonly_resolver_output(
+    register_identity: Any, cfg: Any, key: str
+) -> None:
+    cfg = OmegaConf.create(cfg)
+    sub_key: Any
+    parent_key, sub_key = key.rsplit(".", 1)
+    try:
+        sub_key = int(sub_key)  # convert list index to integer
+    except ValueError:
+        pass
+    parent_node = OmegaConf.select(cfg, parent_key)
+    with pytest.raises(ReadonlyConfigError):
+        parent_node[sub_key] = -1
+
+
+def test_interpolation_readonly_node() -> None:
+    cfg = OmegaConf.structured(User(name="7", age=II("name")))
+    resolved = cfg._get_node("age")._dereference_node()
+    assert resolved == 7
+    with pytest.raises(ReadonlyConfigError):
+        resolved._set_value(8)
```

Comment:
Check the readonly flag directly here and in other places:
```suggestion
    assert resolved._get_flag("readonly") is True
```

Reply:
Done in fefffc1 (without `is True` though since it seems redundant, considering that flags are `Optional[bool]`)